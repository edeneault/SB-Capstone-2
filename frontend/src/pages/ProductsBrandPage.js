import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Row, Col, Container } from "react-bootstrap";
import Product from "../components/Product";
import Loader from "../components/Loader";
import Paginate from "../components/Paginate";

import BrandCarousel from "../components/BrandCarousel";
import { listBrandProducts } from "../actions/productActions";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const ProductsBrandScreen = ({ match }) => {
  const brand = match.params.brand;

  const pageNumber = match.params.pageNumber || 1;
  const dispatch = useDispatch();

  const productBrand = useSelector((state) => state.productBrand);
  const { loading, error, products, page, pages } = productBrand;

  useEffect(() => {
    dispatch(listBrandProducts(brand, pageNumber));
  }, [dispatch, brand, pageNumber]);

  return (
    <>
      <BrandCarousel brand={brand} />
      <Container fluid className='px-5 pt-3'>
        {loading ? (
          <Loader />
        ) : error ? (
          toast(`${error}`, { type: "error" })
        ) : (
          <>
            <Row>
              {products.map((product) => (
                <Col sm={12} md={6} lg={4} xl={3} key={product._id}>
                  <Product product={product} />
                </Col>
              ))}
            </Row>
            <Paginate pages={pages} page={page} brand={brand ? brand : ""} />
          </>
        )}
      </Container>
    </>
  );
};

export default ProductsBrandScreen;
