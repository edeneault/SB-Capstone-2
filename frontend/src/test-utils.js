import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
