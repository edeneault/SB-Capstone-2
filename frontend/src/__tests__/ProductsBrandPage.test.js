import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { initialState, render } from "../test-utils";
import ProductsBrandPage from "../pages/ProductsBrandPage.js";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  let brand;
  beforeEach(() => {
    jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
    brand = "lavazza";
  });

  afterEach(() => {
    jest.spyOn(global.Math, "random").mockRestore();
  });
  test("renders ProductsBrandPage without crashing", async () => {
    const store = mockStore(initialState);
    const match = {
      params: {
        brand: brand,
      },
    };

    const { getByText } = render(
      <Provider store={store}>
        <ProductsBrandPage match={match} />
      </Provider>,
      { initialRoutes: [`/allproducts`] },
    );

    expect(getByText(/~ BRAND LAVAZZA ~/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);
    const match = {
      params: {
        keyword: "",
      },
    };

    const { asFragment } = render(
      <Provider store={store}>
        <ProductsBrandPage match={match} />
      </Provider>,
      { initialRoutes: [`/allproducts`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
