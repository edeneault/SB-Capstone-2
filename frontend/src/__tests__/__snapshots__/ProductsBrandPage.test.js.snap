// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="carousel-title m-0 p-3 text-dark fs-1"
  >
    ~ Brand  ~
  </h2>
  <div
    class="react-multi-carousel-list  fade-in"
  >
    <ul
      class="react-multi-carousel-track "
      style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
    />
  </div>
  <div
    class="px-5 pt-3 container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de3b"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630782239/eShop/tz9ndkjukkrwdneon2fi.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de3b"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Bremille - Barista Express - Red
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star-half-alt"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       2 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $799.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/613515b99f3902000498da5e"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/c_fill,h_460,w_510/v1630868953/eShop/wwcesypqxw4ntj3a78ac.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/613515b99f3902000498da5e"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Bremille - Coffee Shop Barista Red
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       0 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $1499.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de3d"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630782152/eShop/sfpm0zpmddeuarcaofxj.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de3d"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Bremille - Barista Pro
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       12 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $1499.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6133c7430733da86bd484034"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630783375/eShop/hv1lqt49gjevdnrc6zma.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6133c7430733da86bd484034"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Bremille - Coffee Shop Barista
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       2 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $1499.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
